grove/plugin/jwt_auth.go

```diff
@@ -0,0 +1,162 @@
+package plugin
+
+/*
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	log "github.com/apache/trafficcontrol/lib/go-log"
+	"github.com/dgrijalva/jwt-go"
+	"github.com/dgrijalva/jwt-go/request"
+	"github.com/iancoleman/strcase"
+)
+
+type JwtSecret struct {
+	Issuer   string `json:"iss"`
+	Audience string `json:"aud"`
+	Algoritm string `json:"alg"`
+	Kid      string `json:"kid"`
+	Value    string `json:"value"`
+}
+
+type JwtConfig struct {
+	ExportClaims []string    `json:"export_claims"`
+	Secrets      []JwtSecret `json:"secrets"`
+}
+
+type JwtContext struct {
+	SecretsIndex map[string]interface{}
+}
+
+func init() {
+	AddPlugin(10000, Funcs{onRequest: jwtAuth, startup: jwtStartup, load: jwtLoad})
+}
+
+func jwtAuth(icfg interface{}, d OnRequestData) bool {
+	if d.R.Method == http.MethodOptions {
+		return false
+	}
+	jwtContext := (*d.Context).(JwtContext)
+	jwtConfig := icfg.(*JwtConfig)
+	tokenOnRequest, err := request.OAuth2Extractor.ExtractToken(d.R)
+	if err != nil {
+		d.W.WriteHeader(401)
+		d.W.Write([]byte(err.Error()))
+		return true
+	}
+	parser := &jwt.Parser{}
+	token, err := parser.ParseWithClaims(tokenOnRequest, &jwt.MapClaims{}, func(token *jwt.Token) (interface{}, error) {
+		claims := token.Claims.(*jwt.MapClaims)
+		var kid, iss, aud string
+		if value, ok := (*claims)["kid"].(string); ok {
+			kid = value
+		}
+		if value, ok := (*claims)["iss"].(string); ok {
+			iss = value
+		}
+		if value, ok := (*claims)["aud"].(string); ok {
+			aud = value
+		}
+
+		if secret := jwtContext.SecretsIndex[getSecretIndexKey(kid, iss, aud)]; secret != nil {
+			return secret, nil
+		}
+		return nil, fmt.Errorf("Unknown key")
+	})
+	if err != nil {
+		d.W.WriteHeader(401)
+		d.W.Write([]byte(err.Error()))
+		return true
+	}
+	if !token.Valid {
+		d.W.WriteHeader(401)
+		d.W.Write([]byte("Invalid token"))
+		return true
+	}
+	if claims, ok := token.Claims.(*jwt.MapClaims); ok {
+		for _, claimName := range jwtConfig.ExportClaims {
+			if value := (*claims)[claimName]; value != nil {
+				if valueString, ok := value.(string); ok {
+					d.R.Header.Set(fmt.Sprintf("X-Claim-%s", strcase.ToCamel(claimName)), valueString)
+				}
+			}
+		}
+	}
+	return false
+}
+
+func getSecretIndexKey(kid, iss, aud string) string {
+	if kid != "" {
+		return kid
+	}
+	s := ""
+	if iss != "" {
+		s += iss
+	}
+	if aud != "" {
+		s += ":" + aud
```

Comment:
Yes. Though if we're here, we're already in a fairly strange place, because there really should be a `kid` under typical circumstances and both the `iss` and `aud` claims are mandatory. This is a stopgap for finding the right key when no `kid` is available. Obviously, a `kid` is strongly advised because otherwise, key-rolling is impossible. But you don't always get to pick your clients.